### ISSUE TITLE
batchlog_manager_test fails with assert()

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#716]
scylla HEAD: `0687d74`
seastar HEAD: `51154f7`
Setup: Fedora 23
Compiler: `g++ (GCC) 5.3.1 20151207 (Red Hat 5.3.1-2)`
```
 ⋯  urchin   0687d74 … 34 ⚑ 1  build/release/tests/batchlog_manager_test --collectd 0                           
Running 1 test case...
Start Storage service ...
batchlog_manager_test: /home/vladz/work/urchin/seastar/core/sharded.hh:439: Service& seastar::sharded<Service>::local() [with Service = database]: Assertion `local_is_initialized()' failed.
unknown location(0): fatal error in "test_execute_batch": signal: SIGABRT (application abort requested)
seastar/tests/test-utils.cc(31): last checkpoint

*** 1 failure detected in test suite "tests/batchlog_manager_test.cc"
batchlog_manager_test: /home/vladz/work/urchin/seastar/core/sharded.hh:327: seastar::sharded<Service>::~sharded() [with Service = locator::snitch_ptr]: Assertion `_instances.empty()' failed.
Aborted (core dumped)
```

**Comment by slivne** [comment]
Some of the other tests fail on the same thing when trying to run them manually

```
cql_query_test: /home/shlomi/scylla/seastar/core/sharded.hh:439: Service& seastar::sharded<Service>::local() [with Service = database]: Assertion `local_is_initialized()' failed.
unknown location(0): fatal error in "test_create_keyspace_statement": signal: SIGABRT (application abort requested)
seastar/tests/test-utils.cc(31): last checkpoint

shlomi@localhost~/scylla (master)$ ./build/release/tests/storage_proxy_test --collectd 0
Running 2 test cases...
Start Storage service ...
storage_proxy_test: /home/shlomi/scylla/seastar/core/sharded.hh:439: Service& seastar::sharded<Service>::local() [with Service = database]: Assertion `local_is_initialized()' failed.
unknown location(0): fatal error in "test_make_local_reader": signal: SIGABRT (application abort requested)
seastar/tests/test-utils.cc(31): last checkpoint
```

**Comment by slivne** [comment]
#715 seems to be related to this as well

**Comment by vladzcloudius** [comment]
On 12/28/2015 01:14 PM, slivne wrote:
>
> #715 <https://github.com/scylladb/scylla/issues/715> seems to be 
> related to this as well
>

In #715 the tests are timed out - here we get an assert. So although 
they may be related you can't be sure these are not different issues. 
Therefore different issues.

**Comment by slivne** [comment]
If you run the test in the foreground you will see that it prints this
output ... as such I think they may be related - yet there may be more to
it if there are multiple bugs.

**Comment by tgrabiec** [comment]
This assert is usually triggered by an exception thrown from initialization routine causing sharded<> services to fall out of scope before being stopped. You can try to catch that exception by running the test under gdb and adding `catch throw` breakpoint.

**Comment by slivne** [comment]
the issue was introduced by 

commit f57ba6902b044ed06db8d83c2eb0ee9acebd125e
Author: Asias He <asias@scylladb.com>
Date:   Fri Dec 25 10:40:31 2015 +0800

    storage_service: Introduce ring_delay_ms option
    
    It is hard-coded as 30 seconds at the moment.
    
    Usage:
    $ scylla --ring-delay-ms 5000
    
    Time a node waits to hear from other nodes before joining the ring in
    milliseconds.
    
    Same as -Dcassandra.ring_delay_ms in cassandra.



```
[Switching to Thread 0x7ffff09ff700 (LWP 3418)]
0x00007ffff3d59a98 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install boost-filesystem-1.58.0-9.fc23.x86_64 boost-program-options-1.58.0-9.fc23.x86_64 boost-system-1.58.0-9.fc23.x86_64 boost-test-1.58.0-9.fc23.x86_64 boost-thread-1.58.0-9.fc23.x86_64 cryptopp-5.6.2-9.fc23.x86_64 gmp-6.0.0-12.fc23.x86_64 gnutls-3.4.7-1.fc23.x86_64 gnutls-c++-3.4.7-1.fc23.x86_64 hwloc-libs-1.11.0-4.fc23.x86_64 jsoncpp-0.6.0-0.18.rc2.fc23.x86_64 keyutils-libs-1.5.9-7.fc23.x86_64 krb5-libs-1.14-4.fc23.x86_64 libaio-0.3.110-5.fc23.x86_64 libcom_err-1.42.13-3.fc23.x86_64 libffi-3.1-8.fc23.x86_64 libgcc-5.3.1-2.fc23.x86_64 libidn-1.32-1.fc23.x86_64 libpciaccess-0.13.4-2.fc23.x86_64 libselinux-2.4-4.fc23.x86_64 libstdc++-5.3.1-2.fc23.x86_64 libtasn1-4.5-2.fc23.x86_64 libtool-ltdl-2.4.6-7.fc23.x86_64 libxml2-2.9.3-2.fc23.x86_64 lz4-r131-1.fc23.x86_64 nettle-3.1.1-6.fc23.x86_64 numactl-libs-2.0.10-3.fc23.x86_64 openssl-libs-1.0.2e-3.fc23.x86_64 p11-kit-0.23.1-4.fc23.x86_64 pcre-8.38-4.fc23.x86_64 snappy-1.1.3-1.fc23.x86_64 thrift-0.9.1-17.fc23.x86_64 xz-libs-5.2.1-3.fc23.x86_64 yaml-cpp-0.5.1-10.fc23.x86_64 zlib-1.2.8-9.fc23.x86_64
(gdb) where
#0  0x00007ffff3d59a98 in raise () from /lib64/libc.so.6
#1  0x00007ffff3d5b69a in abort () from /lib64/libc.so.6
#2  0x00007ffff3d52227 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ffff3d522d2 in __assert_fail () from /lib64/libc.so.6
#4  0x0000000000875ee9 in seastar::sharded<database>::local (this=<optimized out>) at /home/shlomi/scylla/seastar/core/sharded.hh:439
#5  0x0000000000d337dd in service::storage_service::get_ring_delay (this=<optimized out>) at service/storage_service.cc:2629
#6  0x0000000000b45cd6 in gms::gossiper::quarantine_delay () at gms/gossiper.cc:95
#7  0x0000000000b4cdc9 in gms::gossiper::gossiper (this=0x601000355800) at gms/gossiper.cc:110
#8  0x0000000000f06ea8 in shared_ptr_make_helper<gms::gossiper, true>::make<>() () at /home/shlomi/scylla/seastar/core/shared_ptr.hh:491
#9  make_shared<gms::gossiper> () at /home/shlomi/scylla/seastar/core/shared_ptr.hh:501
#10 seastar::sharded<gms::gossiper>::create_local_service<>() (this=0x194f090 <gms::_the_gossiper>) at ./seastar/core/sharded.hh:311
#11 future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()()::{lambda()#1}::operator()() const (__closure=<optimized out>)
    at ./seastar/core/sharded.hh:340
#12 apply_helper<future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()()::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply({lambda()#1}&&, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()()::{lambda()#1}) (args=<optimized out>, 
    func=<optimized out>) at ./seastar/core/apply.hh:34
#13 apply<future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()()::{lambda()#1}>(future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()()::{lambda()#1}&&, std::tuple<>&&) (args=<optimized out>, func=<optimized out>) at ./seastar/core/apply.hh:42
#14 future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}::operator()() (__closure=0x6010002c7500, __closure@entry=0x601000355800)
    at ./seastar/core/sharded.hh:338
#15 do_void_futurize_apply<future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}&> (func=...) at ./seastar/core/future.hh:1138
#16 0x0000000000f1263d in futurize<void>::apply<future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}>(future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}&&) (func=<optimized out>) at ./seastar/core/future.hh:1186
#17 smp::submit_to<future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int)::{lambda()#1}>(unsigned int, futurize&&) (
    func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xdf9d15b, DIE 0xe25feb6>, t=<optimized out>) at ./seastar/core/reactor.hh:926
#18 future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}::operator()(unsigned int) (c=<optimized out>, __closure=<optimized out>) at ./seastar/core/sharded.hh:342
#19 future<> parallel_for_each<boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}>(boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&)::{lambda(parallel_for_each_state&)#1}::operator()(parallel_for_each_state) const (__closure=__closure@entry=0x6010002c75e0, state=...) at /home/shlomi/scylla/seastar/core/future-util.hh:92
#20 0x0000000000f2528e in do_with<parallel_for_each_state, future<> parallel_for_each<boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}>(boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&)::{lambda(parallel_for_each_state&)#1}>(future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&, future<> parallel_for_each<boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}>(boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&)::{lambda(parallel_for_each_state&)#1}) (
    f=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xdf9d15b, DIE 0xe2c0c9c>, 
    rvalue=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xdf9d15b, DIE 0xe2c0d6b>) at /home/shlomi/scylla/seastar/core/do_with.hh:55
#21 parallel_for_each<boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}>(boost::range_detail::integer_iterator<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&) (
    func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xdf9d15b, DIE 0xe2bebdc>, end=..., begin=...) at /home/shlomi/scylla/seastar/core/future-util.hh:111
#22 parallel_for_each<boost::integer_range<unsigned int>, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}>(boost::integer_range<unsigned int>&&, future<> seastar::sharded<gms::gossiper>::start<>()::{lambda(unsigned int)#1}&&) (func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xdf9d15b, DIE 0xe1fc289>, range=<optimized out>)
    at /home/shlomi/scylla/seastar/core/future-util.hh:133
#23 seastar::sharded<gms::gossiper>::start<>() (this=0x194f090 <gms::_the_gossiper>) at ./seastar/core/sharded.hh:351
#24 0x0000000000f49270 in <lambda()>::operator()(void) const (__closure=__closure@entry=0x6010002c7930) at tests/cql_test_env.cc:85
#25 0x0000000000f4a8fd in apply_helper<tst_init_ms_fd_gossiper(sstring, uint16_t, db::seed_provider_type, sstring)::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (
    args=<optimized out>, func=<optimized out>) at ./seastar/core/apply.hh:34
#26 apply<tst_init_ms_fd_gossiper(sstring, uint16_t, db::seed_provider_type, sstring)::<lambda()> > (args=<optimized out>, func=<optimized out>) at ./seastar/core/apply.hh:42
#27 futurize<future<> >::apply<tst_init_ms_fd_gossiper(sstring, uint16_t, db::seed_provider_type, sstring)::<lambda()> > (args=<optimized out>, 
    func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xe9037b9>) at ./seastar/core/future.hh:1193
#28 future<>::then<tst_init_ms_fd_gossiper(sstring, uint16_t, db::seed_provider_type, sstring)::<lambda()>, future<> > (
    func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xe91c947>, this=0x6010002c7880) at ./seastar/core/future.hh:832
---Type <return> to continue, or q <return> to quit---
#29 tst_init_ms_fd_gossiper (port=7000, cluster_name=<error reading variable: access outside bounds of object referenced via synthetic pointer>, seed_provider=..., 
    listen_address=<error reading variable: access outside bounds of object referenced via synthetic pointer>) at tests/cql_test_env.cc:91
#30 <lambda()>::operator() (__closure=<optimized out>) at tests/cql_test_env.cc:103
#31 apply_helper<init_once(shared_ptr<seastar::sharded<database> >)::<lambda()>, std::tuple<>&&, std::integer_sequence<long unsigned int> >::apply (func=<optimized out>, args=<optimized out>)
    at ./seastar/core/apply.hh:34
#32 apply<init_once(shared_ptr<seastar::sharded<database> >)::<lambda()> > (func=<optimized out>, args=<optimized out>) at ./seastar/core/apply.hh:42
#33 futurize<future<> >::apply<init_once(shared_ptr<seastar::sharded<database> >)::<lambda()> >(<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xe9038f2>, <unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xe9038df>) (args=<optimized out>, func=<optimized out>) at ./seastar/core/future.hh:1193
#34 0x0000000000f4c64a in future<>::then<init_once(shared_ptr<seastar::sharded<database> >)::<lambda()>, future<> > (func=<optimized out>, this=0x6010002c7ac0) at ./seastar/core/future.hh:832
#35 init_once (db=...) at tests/cql_test_env.cc:101
#36 0x0000000000fbcfc3 in single_node_cql_env::start()::{lambda()#1}::operator()() const (__closure=__closure@entry=0x60100005ff18) at tests/cql_test_env.cc:290
#37 0x0000000000fbf1d2 in apply_helper<single_node_cql_env::start()::{lambda()#1}, std::tuple<>&&, std::integer_sequence<unsigned long> >::apply({lambda()#1}&&, std::tuple<>) (args=<optimized out>, 
    func=<optimized out>) at ./seastar/core/apply.hh:34
#38 apply<single_node_cql_env::start()::{lambda()#1}>(single_node_cql_env::start()::{lambda()#1}&&, std::tuple<>&&) (args=<optimized out>, func=<optimized out>) at ./seastar/core/apply.hh:42
#39 do_void_futurize_apply_tuple<single_node_cql_env::start()::{lambda()#1}>(std::result_of&&, std::tuple<(single_node_cql_env::start()::{lambda()#1})...>&&) (
    args=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xebaae66>, 
    func=<unknown type in /home/shlomi/scylla/build/release/tests/batchlog_manager_test_g, CU 0xe52c363, DIE 0xebaae53>) at ./seastar/core/future.hh:1161
#40 futurize<void>::apply<single_node_cql_env::start()::{lambda()#1}>(single_node_cql_env::start()::{lambda()#1}&&, std::tuple<>&&) (args=<optimized out>, func=<optimized out>) at ./seastar/core/future.hh:1181
#41 futurize<std::result_of<std::decay<single_node_cql_env::start()::{lambda()#1}>::type ()>::type>::type seastar::async<single_node_cql_env::start()::{lambda()#1}>(std::decay&&, (std::decay<single_node_cql_env::start()::{lambda()#1}>::type&&)...)::{lambda(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, std::decay<{lambda()#1}>::type&&)::work&)#1}::operator()(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, std::decay<{lambda()#1}>::type&&)::work)::{lambda()#1}::operator()() const (__closure=<optimized out>) at ./seastar/core/thread.hh:258
#42 std::_Function_handler<void (), futurize<std::result_of<std::decay<single_node_cql_env::start()::{lambda()#1}>::type ()>::type>::type seastar::async<single_node_cql_env::start()::{lambda()#1}>(std::decay&&, (std::decay<single_node_cql_env::start()::{lambda()#1}>::type&&)...)::{lambda(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, std::decay<{lambda()#1}>::type&&)::work&)#1}::operator()(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type> seastar::async<{lambda()#1}>(futurize<std::result_of<std::decay<{lambda()#1}>::type ()>::type>::type, std::decay<{lambda()#1}>::type&&)::work)::{lambda()#1}>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/include/c++/5.3.1/functional:1871
#43 0x000000000041655e in std::function<void ()>::operator()() const (this=0x6010000db390) at /usr/include/c++/5.3.1/functional:2271
#44 seastar::thread_context::main (this=0x6010000db380) at core/thread.cc:139
#45 0x00000000004d5c92 in seastar::thread_context::s_main (lo=<optimized out>, hi=<optimized out>) at core/thread.cc:130
#46 0x00007ffff3d6df30 in ?? () from /lib64/libc.so.6
#47 0x0000000000000000 in ?? ()
```

**Comment by asias** [comment]
The left depends on right one.
gossip -> messaging_service  -> db -> cfg

In  tests/cql_test_env.cc, init_once starts gossip before the db is initialized. 

  0     future<> start() {
  1         bool old_active = false;
  2         if (!active.compare_exchange_strong(old_active, true)) {
  3             throw std::runtime_error("Starting more than one cql_test_env at a time not supported "
  4                                      "due to singletons.");
  5         }
  6         return seastar::async([this] {
  7             utils::fb_utilities::set_broadcast_address(gms::inet_address("localhost"));
  8             utils::fb_utilities::set_broadcast_rpc_address(gms::inet_address("localhost"));
  9             locator::i_endpoint_snitch::create_snitch("SimpleSnitch").get();
 10             auto db = ::make_shared<distributed<database>>();
 11             init_once(db).get();
 12             auto cfg = make_lw_shared<db::config>();
 13             _data_dir = make_lw_shared<tmpdir>();
 14             cfg->data_file_directories() = { _data_dir->path };
 15             cfg->commitlog_directory() = _data_dir->path + "/commitlog.dir";
 16             boost::filesystem::create_directories((_data_dir->path + "/system").c_str());
 17             boost::filesystem::create_directories(cfg->commitlog_directory().c_str());
 18             db->start(std::move(*cfg)).get();
 19             db->invoke_on_all([this] (database& db) {
 20                 return db.init_system_keyspace();
 21             }).get();

Tomek, can we move init_once after line 18?

**Comment by tgrabiec** [comment]
@asias We probably can, but it probably won't solve the problem. The db is stopped after each test case, so if you're relying on it being active, you need to restart gossiper as well.

**Comment by asias** [comment]
@tgrabie the tests/cql_test_env.cc is a mess in the area of initialization. I am not sure how we could fix it.  Can we drop the whole init_once logic and in each test we really start what we should start.

**Comment by tgrabiec** [comment]
@asias The intent of cql_test_env is to start a fresh instance of query_processor and its dependencies in each test, that's handled by start() and stop(). init_once() obviously doesn't fit this model because it starts things only once - I am guessing that someone added it as a workaround for the fact that some dependencies didn't have stop interface. I think the right direction is to add stop interface to those and get rid of init_once.

**Comment by penberg** [comment]
The test works fine here after @asias's fixes:

```
[penberg@nero scylla]$ ./build/release/tests/batchlog_manager_test 
Running 1 test case...

*** No errors detected
```

Can we close the issue?

**Comment by asias** [comment]
On Mon, Jan 4, 2016 at 9:01 PM, Pekka Enberg <notifications@github.com>
wrote:

> The test works fine here after @asias <https://github.com/asias>'s fixes:
>
> [penberg@nero scylla]$ ./build/release/tests/batchlog_manager_test
> Running 1 test case...
>
> *** No errors detected
>
> Can we close the issue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scylladb/scylla/issues/716#issuecomment-168672235>.
>

After my fix, all the tests are passing.


-- 
Asias
